src/codegen/sys/ffi_type.rs

```diff
@@ -0,0 +1,68 @@
+use analysis::rust_type::Result;
+use env::Env;
+use gobjects::GStatus;
+use library;
+use nameutil::module_name;
+
+pub fn ffi_type(env: &Env, type_id: library::TypeId) -> Result {
+    use library::Type::*;
+    use library::Fundamental::*;
+    let type_ = env.library.type_(type_id);
+    match type_ {
+        &Fundamental(fund) => {
+            let ok = |s: &str| Ok(s.into());
+            let err = |s: &str| Err(s.into());
+            match fund {
+                None => err("()"),
+                Boolean => ok("gboolean"),
+                Int8 => ok("i8"),
+                UInt8 => ok("u8"),
+                Int16 => ok("i16"),
+                UInt16 => ok("u16"),
+                Int32 => ok("i32"),
+                UInt32 => ok("u32"),
+                Int64 => ok("i64"),
+                UInt64 => ok("u64"),
+
+                Int => ok("c_int"),      //maybe dependent on target system
+                UInt => ok("c_uint"),     //maybe dependent on target system
+
+                Float => ok("c_float"),
+                Double => ok("c_double"),
+
+                Utf8 => ok("*const c_char"),
+
+                Type => ok("GType"),
+                Unsupported => err("Unsupported"),
+                _ => err(&format!("Fundamental: {:?}", fund)),
+            }
+        },
+
+        &Enumeration(ref enum_) => Ok(format!("enums::{}", enum_.name)),
+        &Interface(ref interface) => to_mut_ptr(fix_external_name(env, type_id, &interface.glib_type_name)),
+        &Class(ref klass) => to_mut_ptr(fix_external_name(env, type_id, &klass.glib_type_name)),
+        _ => Err(format!("Unknown rust type: {:?}", type_.get_name() )),
+        //TODO: check usage library::Type::get_name() when no _ in this
+    }
+}
+
+fn fix_external_name(env: &Env, type_id: library::TypeId, name: &str) -> Result {
+    if type_id.ns_id == library::MAIN_NAMESPACE || type_id.ns_id == library::INTERNAL_NAMESPACE {
+        Ok(name.into())
+    } else {
+        let name_with_prefix = format!("{}_ffi::{}",
```

Comment:
But we'll need a map of namespaces to prefixes because out `gdk` crate contains both `Gdk` and `GdkPixbuf` and the `glib` crate also has `gio` and `gobject`.